### PR TITLE
Auto detect predicate type and use single parameter for passing it

### LINKED ISSUE #14: predicate and predicate_generator parameters are confusing
We can use a single parameter and auto detect its type.

### PR DESCRIPTION
**Pull request by krystian8207**
### Link to backlog task in asana (if there is no task pls create it!)

Closes #14

Before submitting a PR please answer:
* Was it necessary to update README? Is README updated? yes
* Are unit tests added for new logic? yes

### CODE REVIEW

File: R/assertions.R [hunk 1]
```diff
@@ -13,6 +13,14 @@ validate <- function(data, name, description = NULL) {
   data
 }
 
+get_assert_method <- function(predicate, methods) {
```
Comment (dokato) [inline R/assertions.R:16]: Short description above, pelase

File: R/utils.R [hunk 2]
```diff
@@ -21,3 +21,19 @@ generate_id <- function () {
   paste0(paste0(sample(c(LETTERS, letters, 0:9), 5, TRUE),
                 collapse = ""), round(as.numeric(Sys.time()) * 1000))
 }
+
+#' Detect type of predicate
+#' @return String informing about predicate type.
+#'   When "direct" the predicate is directly applied to the column.
+#'   When "generator" the predicate returns predicate computed based on a column.
+#'   When NA, the predicate shouldn't be used at all in the validation.
+get_predicate_type <- function(predicate) {
```
Comment (dokato) [inline R/utils.R:30]: `@param predicate ...`

**Review by dokato — APPROVED** [review]
🎉 Could you fix these two things before merge?

File: R/utils.R (hunk 2 above)
Comment (krystian8207) [reply R/utils.R:33]: done

File: R/assertions.R (hunk 1 above)
Comment (krystian8207) [reply R/assertions.R:20]: done

File: R/utils.R [hunk 3]
```diff
@@ -16,8 +16,27 @@ get_first_name <- function(df){
...
+  predicate_output <- predicate(0:1)
+  if (is.logical(predicate_output)) {
+    return("direct")
+  }
+  if (is.function(predicate_output)) {
```
Comment (przytu1) [inline R/utils.R:38]: Can we also check if `is.logical(predicate_output(0:1))`?

File: R/assertions.R [hunk 4]
```diff
@@ -27,16 +39,15 @@ assert_if <- function(data, expr, description = NA, obligatory = FALSE, skip_cha
...
 
 #' @export
-assert_cols <- function(data, ..., predicate = NULL, predicate_generator = NULL, obligatory = FALSE, description = NA,
+assert_cols <- function(data, predicate, ..., obligatory = FALSE, description = NA,
                         skip_chain_opts = FALSE, success_fun = assertr::success_append, error_fun = assertr::error_append,
                         defect_fun = assertr::defect_append) {
 
-  assertr_function <- if (!is.null(predicate_generator)) assertr::insist else assertr::assert
-  predicate_arg <- if (!is.null(predicate_generator)) predicate_generator else predicate
+  assertr_function <- get_assert_method(predicate, list(direct = assertr::assert, generator = assertr::insist))
```
Comment (przytu1) [inline R/assertions.R:46]: Do we really need to provide `methods` argument? Are there any other predicate types possible? Maybe this is a premature optimization that causes confusion. I think we just need a function like `get_predicate_type`, for example:

```
get_assert_method <- function(predicate) {
  predicate_output <- predicate(0:1)
  if (is.logical(predicate_output)) {
    return(assertr::assert)
  } else if (is.function(predicate_output) && is.logical(predicate_output(0:1))) {
    return(assertr::insist)
  }  else {
    stop("Predicate should be a function returning logical vector or function")
  }
}
```

WDYT?

**Review by przytu1 — CHANGES_REQUESTED** [review]
Awesome work :100: I had some comments, please take a look

File: R/assertions.R (hunk 4 above)
Comment (krystian8207) [reply R/assertions.R:46]: Your example covers only 50% of cases. I've added this because we also have assert_rows and insist_rows case.

Comment (przytu1) [reply R/assertions.R:46]: I see! I missed that this function is also used in `assert_rows` with different methods. So, if we assume anyway that there is "direct" and "generator" method only, can we have a single function like this:

```
get_assert_method <- function(predicate, method = list(direct = assertr::assert, generator = assertr::insist)) {
  predicate_output <- predicate(0:1)
  if (is.logical(predicate_output)) {
    return(method$direct)
  } else if (is.function(predicate_output) && is.logical(predicate_output(0:1))) {
    return(method$generator)
  }  else {
    stop("Predicate should be a function returning logical vector or function")
  }
}
```

Having such default argument for `methods` in function interface clarifies me a lot. WDYT now?

**Review by przytu1 — CHANGES_REQUESTED** [review]

File: R/assertions.R (hunk 4 above)
Comment (krystian8207) [reply R/assertions.R]: fixed

File: R/utils.R (hunk 3 above)
Comment (krystian8207) [reply R/utils.R]: done

**Comment by przytu1** [comment]
LGTM, great work

**Pull request merged**